### PR TITLE
[Clustering] Fixing the bug in clustering for Conv1D/Conv3D

### PR DESCRIPTION
**Pull request by wwwind**
There is a bug in clustering for Conv1D/Conv3D. We used the layout for Conv2D and this is wrong, because tensors for Conv1D/ Conv3D have different ranks. Tests for both cases are added.

### CODE REVIEW

**Comment by akarmi** [comment]
Thanks. Looks like we need a similar test and fix for Conv1D.

**Review by akarmi — CHANGES_REQUESTED** [review]
Thanks. Looks like we need a similar test and fix for Conv1D.

**Review by akarmi — APPROVED** [review]

**Comment by akarmi** [comment]
@psunn, please review

**Comment by psunn** [comment]
LGTM, thanks!

File: tensorflow_model_optimization/python/core/clustering/keras/cluster_test.py [hunk 1]
```diff
@@ -199,6 +202,54 @@ def testDepthwiseConv2DLayerNonClusterable(self):
                                    wrapped_layer)
     self.assertEqual([], wrapped_layer.layer.get_clusterable_weights())
 
+  @keras_parameterized.run_all_keras_modes
+  def testConv1DLayer(self):
+     """
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/clustering/keras/cluster_test.py:207]: Nit: """Verifies that we can cluster a Conv1D layer.""" Should be a one-liner

File: tensorflow_model_optimization/python/core/clustering/keras/cluster_test.py [hunk 2]
```diff
@@ -199,6 +202,54 @@ def testDepthwiseConv2DLayerNonClusterable(self):
...
+                                    wrapped_layer)
+     self.assertEqual([4, 5, 1, 3],
+      wrapped_layer.layer.get_clusterable_weights()[0][1].shape)
+
+  @keras_parameterized.run_all_keras_modes
+  def testConv3DLayer(self):
+     """
+     Verifies that we can cluster a Conv3D layer.
+     """
+     input_shape =(4, 28, 28, 28, 1)
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/clustering/keras/cluster_test.py:242]: Same as above:

File: tensorflow_model_optimization/python/core/clustering/keras/clustering_registry.py [hunk 3]
```diff
@@ -27,21 +27,22 @@ class ConvolutionalWeightsCA(AbstractClusteringAlgorithm):
...
     tiled_cluster_centroids = tf.stack(
         [tf.tile(tf.stack(
-            [tf.reshape(self.cluster_centroids, [1, 1, clst_num])] *
-            weight.shape[-2], axis=2),
-                 [weight.shape[0], weight.shape[1], 1, 1])] * weight.shape[-1],
-        axis=3)
+            [tf.reshape(self.cluster_centroids, [1 for _ in range(wt_dim-2)] + [clst_num])] *
+            weight.shape[-2], axis=wt_dim-2),
+                 [i for i in weight.shape[:-2]] + [1, 1])] * weight.shape[-1],
+        axis=wt_dim-1)
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/clustering/keras/clustering_registry.py:40]: Seems a little clunky with all the list comprehensions, is it possible to make this a bit more readable?

**Comment by wwwind** [comment]
Hi @daverim Thanks for the review! Your comments are addressed. Please take a look.

**Comment by daverim** [comment]
Hey sorry, is this PR still required? Can you rebase?

**Comment by wwwind** [comment]
@googlebot I signed it!

**Comment by psunn** [comment]
@googlebot I consent

**Comment by wwwind** [comment]
@googlebot I consent.

**Review by akarmi — APPROVED** [review]

**Review by daverim — APPROVED** [review]

**Pull request merged**